### DISCUSSION TITLE
[optimistic update]: cache not updated with mutation for pagination query

### DISCUSSION THREAD

**Discussion by Hugo2121** [urql-graphql/urql#2230] [Q&A] [answered]
I use `urql` to link with `Django graphQL` backend. The backend side has set up Relay Pagination and a simple query would look like:

```
query AdminUsersQuery {
  adminUsers (first: 2, after: "") {
    totalCount
    pageInfo {
      hasNextPage
      hasPreviousPage
      endCursor
    }
    edges {
      __typename
      node {
        __typename
        email
        name
        userType
        supportImageType
      }
      cursor
    }
  }
}
```

the result looks like:

```
{
  "data": {
    "adminUsers": {
      "totalCount": 22,
      "pageInfo": {
        "hasNextPage": true,
        "hasPreviousPage": false,
        "endCursor": "YXJyYXljb25uZWN0aW9uOjE="
      },
      "edges": [
        {
          "__typename": "AdminUserNodeEdge",
          "node": {
            "__typename": "AdminUserNode",
            "email": "asdfzxcvzxcv@sdfxzvc.com",
            "name": "3",
            "userType": 4,
            "supportImageType": [
              "MG",
              "AZ",
              "SA",
              "SR",
              "AZ-GC"
            ]
          },
          "cursor": "YXJyYXljb25uZWN0aW9uOjA="
        },
        {
          "__typename": "AdminUserNodeEdge",
          "node": {
            "__typename": "AdminUserNode",
            "email": "asdfasdf@sdfsdf.com",
            "name": "4",
            "userType": 1,
            "supportImageType": [
              "MG",
              "AZ",
              "SA",
            ]
          },
          "cursor": "YXJyYXljb25uZWN0aW9uOjE="
        }
      ]
    }
  }
}

```

The queried results are showing in frontend correctly. The `graphQL client` set up looks:

```
import {
  createClient, Provider as URQLProvider,
  fetchExchange,
  dedupExchange,
} from 'urql';
import { cacheExchange } from '@urql/exchange-graphcache';
import { relayPagination } from '@urql/exchange-graphcache/extras';
import { devtoolsExchange } from '@urql/devtools';

const client = createClient({
  url: 'graphql',
  requestPolicy: 'cache-and-network',
  exchanges: [
    devtoolsExchange,
    dedupExchange,
    cacheExchange({
      keys: {
        AdminUserNode: (data) => {
          console.log('data is', data);
          return data.email;
        },
      },
      resolvers: {
        Query: {
          adminUsers: relayPagination(),
        },
      },
      optimistic: {
        addNewAdminUser: (variables, cache, info) => (
          {
            __typename: 'AdminUserMutationPayload',
            id: variables?.input?.email,
            ...variables?.input,
          }),
      },
    }),
    fetchExchange,
  ],
});

```

so I have a customer key - email. and my mutation looks like:

```
mutation AdminUsersMutation ($data: AdminUserMutationInput!) {
    addNewAdminUser(input: $data) {
        email
        name
        userType
        supportImageType
    }
}

const [, updateUser] = useMutation(AdminUsersMutation);
updateUser({ data: {email: "test@abc.com", name: "testuser", userType: "2", supportImageType: ["a", "b"]}}).then(result => console.log(result))

```

I can find the updated result in console log, but the rendered result is not updated.

I am not sure if I set up optimistic cache correctly or not,  when I update admin user record with mutation, the update is only happen on server side, the **cache is not updated**, if I refresh the page, the new result will show. 

- How to set up optimistic cache correctly? I found it is a little bit hard to know the exact `__typename` when referring to documentation of `optimistic cache` section
- How to debug cache? I didn't see any error during my mutation, but the cache seems not updated

**Comment by kitten** [comment, marked as answer]
The cache is updated, however, if your mutation is called `addNewAdminUser` then I assume it creates a new entity in your database.
So, thinking about this from first principles, how does your config now tell the normalised cache where this entity is supposed to go?

This section in the docs onwards deals with that question: https://formidable.com/open-source/urql/docs/graphcache/cache-updates/#updating-lists-or-links

Basically, it's easy to miss with a simple CRUD operation, but Graphcache does not know whether your list is sorted, where that list is, what's in that list, or what goes in that list, so you'll need to write an updater that takes care of that

**Reply by Hugo2121** [reply]
sorry about the confusing name: ` addNewAdminUser`, I am actually update the record, for example, change the name from 'a' to 'b', the mutation call is success and DB has been updated, only the page is not re-rendered. To me, it seems the cache is not updated after mutation. So I want to know how to debug or how should I update the cache?

**Reply by JoviDeCroock** [reply]
Well the `__typename` doe not seem to match `AdminUserMutationPayload` is not a `AdminUserNode` which means the cache won't be able to reconcile your entities

**Reply by Hugo2121** [reply]
I also feel the `__typename` may not be correct, to my understanding, the `__typename` act like a key for the caching. I need the correct `__typename` for the whole process to work.

then what should be the correct `__typename`? 
I tried `AdminUserNode` as __typename, also not working

**Reply by Hugo2121** [reply]
some updates so far, 
when updating record with [manually update from doc](https://formidable.com/open-source/urql/docs/graphcache/cache-updates/#manually-updating-entities): 

```
updates: {
        Mutation: {
          addNewAdminUser(_result, args, cache, _info) {
            const fragment = gql`
              fragment _ on AdminUserNode {
                email
                name
                userType
                supportImageType
              }
            `;
            cache.writeFragment(fragment, { ...args?.input });
          },
        },
      },

```

the cache is updated as I can see the result change directly without refresh the page

**Reply by nandorojo** [reply]
I think you just need the mutation's `__typename` to match the query's typename that's already in your cache.
